testing_and_setup/compass/manage_regression_suite.py

```diff
@@ -193,19 +194,242 @@ def process_test_clean(test_tag, work_dir):  # {{{
+    local_parallel_code += "compleated =[]\n"
+    local_parallel_code += "running = []\n"
+    local_parallel_code += "adding_phase = True\n"
+    local_parallel_code += "processing_phase = False\n"
+    local_parallel_code += "done = False\n"
+    local_parallel_code += "\n"
+    local_parallel_code += "while True:\n"
+    local_parallel_code += "  if len(compleated) < len(testcase_data.keys()):\n"
+    local_parallel_code += "    adding_phase = True\n"
+    local_parallel_code += "  else:\n"
+    local_parallel_code += "    adding_phase = False\n"
+    local_parallel_code += "    done = True\n"
+    local_parallel_code += "  if adding_phase:\n"
+    local_parallel_code += "    running_names = [running_proc[1] for running_proc in running]\n"
+    local_parallel_code += "    for key in testcase_data.keys():\n"
+    local_parallel_code += "      if not key in compleated and not key in running_names:\n"
+    local_parallel_code += "        if not testcase_data[key]['prereqs'] == [] and testcase_data[key]['runnable'] == False:\n"
+    local_parallel_code += "          prereq_tests = [prereq['name'] for prereq in testcase_data[key]['prereqs']]\n"
+    local_parallel_code += "          prereq_compleated =  all(prereq in compleated for prereq in prereq_tests)\n"
+    local_parallel_code += "          if prereq_compleated:\n"
+    local_parallel_code += "            testcase_data[key]['runnable'] = True\n"
+    local_parallel_code += "        if number_of_procs >= testcase_data[key]['procs'] and testcase_data[key]['runnable'] == True:\n"
+    local_parallel_code += "          number_of_procs = number_of_procs - testcase_data[key]['procs']\n"
+    local_parallel_code += "          process , key = run(key)\n"
+    local_parallel_code += "          running.append((process,key))\n"
+    local_parallel_code += "        else:\n"
+    local_parallel_code += "          processing_phase = True\n"
+    local_parallel_code += "          adding_phase = False \n"
+    local_parallel_code += "  if processing_phase:\n"
+    local_parallel_code += "    for process in running:\n"
+    local_parallel_code += "      pid = process[0].pid\n"
+    local_parallel_code += "      running_key = process[1]\n"
+    local_parallel_code += "      if not psutil.pid_exists(pid):\n"
+    local_parallel_code += "        running.remove(process)\n"
+    local_parallel_code += "        print('DONE: {} {} {}'.format(testcase_data[key]['procs'], number_of_procs, running_key))\n"
+    local_parallel_code += "        compleated.append(running_key)\n"
+    local_parallel_code += "        number_of_procs = number_of_procs + testcase_data[running_key]['procs']\n"
+    local_parallel_code += "      else:\n"
+    local_parallel_code += "        try:\n"
+    local_parallel_code += "          process[0].wait(timeout=3)\n"
+    local_parallel_code += "        except subprocess.TimeoutExpired:\n"
+    local_parallel_code += "          continue\n"
+    local_parallel_code += "    if running == []:\n"
```

Comment:
This is a bit safer:
```suggestion
    local_parallel_code += "    if len(running) == 0:\n"
```